x-pack/plugins/uptime/public/contexts/uptime_check_steps_context.tsx

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useContext, Context } from 'react';
+import { useParams } from 'react-router-dom';
+import { useCheckSteps } from '../components/synthetics/check_steps/use_check_steps';
+import { JourneyState } from '../state/reducers/journey';
+
+export const CheckStepsContext = React.createContext<JourneyState>({
+  checkGroup: '',
+  steps: [],
+  loading: false,
+});
+
+export const UptimeCheckStepsContextProvider = ({ children }: { children: React.ReactNode }) => {
```

Comment:
What's the reason for this context? Is it to ensure we never issue multiple requests across the entire app, so multiple calls to `useCheckSteps` with the same check group ID will not trigger multiple requests?

If that's the only reason, we might be able to improve on this solution without creating a new context. I think we could cache the results of each check group in the redux store, and check the store before trying to fetch in the hook. This would cover a case where the user clicks "next step" and "previous step" repeatedly, we'd be able to re-render the page much faster. The drawback of this solution is we'd also want to add some delete logic so the locally-stored data doesn't grow unbounded.

LMK if any of my above assumptions are incorrect, or if the alternative solution sounds interesting to you.

Reply:
Hey Justin, this is a really interesting alternative, but I do have one concern about it:

* If we cache the results of the `fetch`, we'd need to determine when to invalidate the cache and request it again as you mention. Both for storage reasons and for "freshness" reasons.
    I'd be concerned that choosing a good heuristic for that could be very difficult as [cache invalidation does tend to be more complex than it seems](https://martinfowler.com/bliki/TwoHardThings.html). Furthermore, IMO it's more complex to store that information within `redux` and create a cache invalidation strategy than it is to use a context.

The main reason I had to create a context here is because I can't lift state up, as these components are passed separately to the function which creates the route components in `routes.ts` and then rendered side-by side. Therefore, the only way to lift up state was to create a wrapper which could be used in `routes.ts`. Now, that could be yet another stateful component, but then it would end up having the same semantics as a context, just without the proper naming.

With a context we can centralize the fetch in this upper component which holds state, and we don't have to worry about cache invalidation.

For the reasons described above, I would consider the context to be simpler than a redux `cache`.

**Please let me know if you think I missed some of the advantages of caching or if I misunderstood your suggestion**.

CC @shahzad31 I'd love to hear your thoughts on this as well.